src/items/AppItem.tsx

```diff
@@ -79,260 +68,109 @@ export interface AppItemProps {
-
-    // responds only to corresponding app
-    if (!url?.includes(requestOrigin)) {
-      return;
-    }
-
-    const { type, payload } = JSON.parse(data);
-
-    switch (type) {
-      case POST_MESSAGE_KEYS.GET_AUTH_TOKEN: {
-        // eslint-disable-next-line no-unused-expressions
-        channel?.port1.postMessage(
-          JSON.stringify({
-            type: POST_MESSAGE_KEYS.GET_AUTH_TOKEN_SUCCESS,
-            payload: {
-              token: await this.getToken(payload),
-            },
-          }),
-        );
-        break;
-      }
-
-      case POST_MESSAGE_KEYS.POST_AUTO_RESIZE: {
-        // item should not be manually resizable
-        if (this.props.isResizable) {
-          return;
-        }
-        // iframe must be mounted
-        if (this.iframeRef.current === null) {
-          return;
-        }
-        // payload should be number
-        if (typeof payload !== 'number') {
-          return;
-        }
-        this.iframeRef.current.height = payload.toString();
-        break;
-      }
-    }
-  };
-
-  windowOnMessage = (e: MessageEvent): void => {
-    const { item, member, memberId, apiHost, context, permission } = this.props;
-    const { url } = this.state;
-    const { data, origin: requestOrigin } = e;
-
-    const POST_MESSAGE_KEYS = buildPostMessageKeys(item.id);
-
-    // responds only to corresponding app
-    if (!url?.includes(requestOrigin)) {
-      return;
-    }
-
-    // return context data and message channel port to app
-    const { type } = JSON.parse(data);
-    if (type === POST_MESSAGE_KEYS.GET_CONTEXT) {
-      // create/reset channel and
-      // Listen for messages on port1
-      const channel = new MessageChannel();
-      const { port1 } = channel;
-      this.setState({ channel });
-      port1.onmessage = this.onMessage;
-
-      // Transfer port2 to the iframe
-      // provide port2 to app and item's data
-      // eslint-disable-next-line no-unused-expressions
-      this.iframeRef?.current?.contentWindow?.postMessage(
-        JSON.stringify({
-          type: POST_MESSAGE_KEYS.GET_CONTEXT_SUCCESS,
-          payload: {
-            apiHost,
-            itemId: item.id,
-            settings: item.settings,
-            memberId: memberId || member?.id,
-            permission,
-            lang: item.settings?.lang || member?.extra?.lang || DEFAULT_LANG,
-            context,
-          },
-        }),
-        '*',
-        [channel.port2],
-      );
-    }
-  };
-
-  render(): JSX.Element {
-    const {
+};
+
+const AppItem = ({
+  id,
+  item,
+  memberId,
+  member,
+  apiHost,
+  context,
+  requestApiAccessToken,
+  height = DEFAULT_APP_HEIGHT,
+  editCaption = false,
+  showCaption = true,
+  permission = PermissionLevel.Read,
+  isResizable = false,
+  showCollapse,
+  onSaveCaption,
+  onCancelCaption,
+  saveButtonId,
+  cancelButtonId,
+}: AppItemProps): JSX.Element => {
+  // state
+  const [isiFrameLoading, setIsiFrameLoading] = useState(true);
```

Comment:
style nitpick (since the DOM element type is `HTMLIFrameElement`)
```suggestion
  const [isIFrameLoading, setIsIFrameLoading] = useState(true);
```